storage/database/db_migration.go

```diff
@@ -75,29 +57,113 @@ loop:
 			return errors.Wrap(err, "failed to put batch")
 		}
 
+		if dstBatch.ValueSize() > IdealBatchSize {
+			if err := dstBatch.Write(); err != nil {
+				return err
+			}
+			dstBatch.Reset()
+		}
+
 		// make a report
-		if fetched%reportCycle == 0 {
+		if fetched%(IdealBatchSize*20) == 0 {
 			logger.Info("DB migrated",
-				"fetched", fetched-previousFetched, "elapsedIter", time.Since(cycleStart),
-				"fetchedTotal", fetched, "elapsedTotal", time.Since(start))
-			cycleStart = time.Now()
-			previousFetched = fetched
+				"db", name, "fetchedTotal", fetched, "elapsedTotal", time.Since(start))
 		}
 
 		// check for quit signal from OS
 		select {
-		case <-sigQuit:
-			logger.Info("exit called", "fetchedTotal", fetched, "elapsedTotal", time.Since(start))
-			break loop
+		case <-quit:
+			logger.Warn("exit called", "db", name, "fetchedTotal", fetched, "elapsedTotal", time.Since(start))
+			return nil
 		default:
 		}
 	}
 
 	if err := dstBatch.Write(); err != nil {
 		return errors.Wrap(err, "failed to write items")
 	}
+	dstBatch.Reset()
+
+	logger.Info("Finish DB migration", "db", name, "fetchedTotal", fetched, "elapsedTotal", time.Since(start))
+
+	srcIter.Release()
+	if err := srcIter.Error(); err != nil { // any accumulated error from iterator
+		return errors.Wrap(err, "failed to iterate")
+	}
+
+	return nil
+}
+
+// StartDBMigration migrates a DB to another DB.
+// (e.g. LevelDB -> LevelDB, LevelDB -> BadgerDB, LevelDB -> DynamoDB)
+// Do not use db migration while a node is executing.
+func (dbm *databaseManager) StartDBMigration(dstdbm DBManager) error {
+	// settings for quit signal from os
+	quit := make(chan struct{})
+	go func() {
+		sigc := make(chan os.Signal, 1)
+		signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM)
+		defer signal.Stop(sigc)
+		<-sigc
+		logger.Info("Got interrupt, shutting down...")
+		close(quit)
+		for i := 10; i > 0; i-- {
+			<-sigc
+			if i > 1 {
+				logger.Info("Already shutting down, interrupt more to panic.", "times", i-1)
+			}
+		}
```

Comment:
What is this for?

Reply:
@winnie-byun This is to gracefully terminate the process. If you send more than 10 signals, it kills the process instantly. Otherwise, it will wait the main goroutine to be terminated.